core/src/impl/Kokkos_Profiling_Interface.cpp

```diff
@@ -52,6 +52,20 @@ namespace Kokkos {
        	return (NULL != initProfileLibrary);
     }
 
+    void pushRegion(const std::string& kName) {
+	if( NULL != pushRegionCallee ) {
+	    Kokkos::fence();
```

Comment:
@hcedwar so yes multiple threads can make these calls but the function pointers are read only. It is up to the tools to decide how to handle thread safety on the callee side.

**Why would we do this?** Because we want to allow thread safe data structures to be used on the tool side or alternatively locks where needed. Each tool has unique requirements and so we want to defer that choice to the implementor.

The approach also prevents the addition of complex overheads in Kokkos which may not be required. The initialization can tell the tools whether threading is being used etc, so some tools can run with zero overhead as well.